test/common/quic/envoy_quic_h3_fuzz_helper.cc

```diff
@@ -0,0 +1,431 @@
+  quic::QuicPathFrameBuffer buffer = {0};
+  size_t to_copy = std::min(data.size(), buffer.size());
+  const char* start = data.data();
+  const char* end = start + to_copy;
+  std::copy(start, end, buffer.begin());
+  return buffer;
+}
+
+static quic::QuicErrorCode toErrorCode(uint32_t) { return quic::QuicErrorCode::QUIC_NO_ERROR; }
+
+QuicPacketizer::QuicPacketizer(const quic::ParsedQuicVersion& quic_version,
+                               quic::QuicConnectionHelperInterface* connection_helper)
+    : quic_version_(quic_version), connection_helper_(connection_helper), packet_number_(0),
+      destination_connection_id_(quic::test::TestConnectionId()),
+      framer_({quic_version_}, connection_helper_->GetClock()->Now(), quic::Perspective::IS_CLIENT,
+              quic::kQuicDefaultConnectionIdLength),
+      h3packetizer_(open_h3_streams_) {
+  quic::Perspective p = quic::Perspective::IS_CLIENT;
+  framer_.SetEncrypter(quic::ENCRYPTION_INITIAL, std::make_unique<FuzzEncrypter>(p));
+  framer_.SetEncrypter(quic::ENCRYPTION_HANDSHAKE, std::make_unique<FuzzEncrypter>(p));
+  framer_.SetEncrypter(quic::ENCRYPTION_ZERO_RTT, std::make_unique<FuzzEncrypter>(p));
+  framer_.SetEncrypter(quic::ENCRYPTION_FORWARD_SECURE, std::make_unique<FuzzEncrypter>(p));
+}
+
+void QuicPacketizer::serializePackets(const QuicH3FuzzCase& input) {
+  for (auto& quic_frame_or_junk : input.frames()) {
+    if (idx_ >= sizeof(quic_packet_sizes_) / sizeof(quic_packet_sizes_[0])) {
+      return;
+    }
+    if (quic_frame_or_junk.has_qframe()) {
+      serializePacket(quic_frame_or_junk.qframe());
+    } else if (quic_frame_or_junk.has_junk()) {
+      const std::string& junk = quic_frame_or_junk.junk();
+      serializeJunkPacket(junk);
+    }
+  }
+}
+
+void QuicPacketizer::foreach (std::function<void(const char*, size_t)> cb) {
+  for (size_t i = 0; i < idx_; i++) {
+    if (quic_packet_sizes_[i] > 0) {
+      cb(quic_packets_[i], quic_packet_sizes_[i]);
+    }
+  }
+}
+
+void QuicPacketizer::reset() {
+  idx_ = 0;
+  packet_number_ = quic::QuicPacketNumber(0);
+  for (unsigned long& quic_packet_size : quic_packet_sizes_) {
+    quic_packet_size = 0;
+  }
+  open_h3_streams_.clear();
+}
+
+void QuicPacketizer::serializePacket(const QuicFrame& frame) {
+  char buffer[1024];
+
+  std::unique_ptr<quic::QuicCryptoFrame> crypto_frame = nullptr;
+  std::unique_ptr<quic::QuicRstStreamFrame> reset_stream_frame = nullptr;
+  std::unique_ptr<quic::QuicConnectionCloseFrame> connection_close_frame = nullptr;
+  std::unique_ptr<quic::QuicGoAwayFrame> go_away_frame = nullptr;
+  std::unique_ptr<quic::QuicRetireConnectionIdFrame> retire_connection_id_frame = nullptr;
+  std::unique_ptr<quic::QuicMessageFrame> message_frame = nullptr;
+  std::unique_ptr<quic::QuicNewTokenFrame> new_token = nullptr;
+  std::unique_ptr<quic::QuicAckFrequencyFrame> ack_frequency = nullptr;
+  quic::QuicAckFrame ack_frame;
+  quic::QuicNewConnectionIdFrame new_connection_id_frame;
```

Comment:
Are these used anywhere?

Reply:
Yes, they are constructed inside the switch cases. See the other comment regarding the switch case.